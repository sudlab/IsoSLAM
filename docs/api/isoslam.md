# IsoSLAM

This module contains the main functions that constitute the IsoSLAM package.

::: isoslam.isoslam
    handler: python
    options:
        docstring_style: numpy
        rendering:
            show_signature_annotations: true
