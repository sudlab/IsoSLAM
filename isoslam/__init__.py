"""IsoSLAM."""

from importlib.metadata import version

release = version("isoslam")
__version__ = ".".join(release.split("."[:2]))
