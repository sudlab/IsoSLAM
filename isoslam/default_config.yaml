# Default configuration file for running IsoSlam
log_level: "warning"
output_dir: output
output_file: results.parquet
bam_file: data/bam/.bam
gtf_file: data/gtf/
bed_file: data/bed/
vcf_file: data/vcf/
upper_pairs_limit: 2000000
first_matched_limit: 1000
forward_reads:
  from: "T"
  to: "C"
reverse_reads:
  from: "A"
  to: "G"
delim: "\t"
schema:
  read_uid: "int"
  transcript_id: "str"
  start: "int"
  end: "int"
  chr: "str"
  strand: "str"
  assignment: "str"
  conversions: "int"
  convertible: "int"
  coverage: "int"
summary_counts:
  file_ext: ".tsv"
  directory: "output"
  groupby:
    - Transcript_id
    - Chr
    - Strand
    - Start
    - End
    - Assignment
    - Conversions
    - filename
  conversions_var: "Conversions"
  conversions_threshold: 1
  test_file: "no4sU"
  filename_var: "filename"
  regex: ^d(\w+)_(\w+)hr(\w+)_
  output:
    outfile: summary_counts.tsv
    sep: "\t"
plot: false
