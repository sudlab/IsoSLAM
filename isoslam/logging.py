"""Setup and configure logging."""

import sys

from loguru import logger


def setup(level: str = "INFO") -> None:
    """
    Loguru setup with the required logging level and format.

    Parameters
    ----------
    level : str
        Log level, default is "INFO", other options "WARNING", "DEBUG" etc.
    """
    logger.remove()
    logger.add(sys.stderr)
    logger.add(
        sys.stderr,
        colorize=True,
        level=level.upper(),
        format="<green>{time:HH:mm:ss}</green> "
        "| <level>{level}</level> | "
        "<magenta>{file}</magenta>:<magenta>{module}</magenta>:<magenta>{function}</magenta>:<magenta>{line}</magenta>"
        " | <level>{message}</level>",
    )
