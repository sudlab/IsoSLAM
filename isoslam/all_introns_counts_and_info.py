"""
all_introns_counts_and_info.py
====================
Takes in the sorted and feature assigned `.bam` file from previous steps and passes them to a
python script that uses pysam (python wrapper for htslib). This script iterates over the `.bam`
file pair-by-pair (representing the sequencing insert), determines whether the read-pair shows
evidence of intron splicing/retention and assigns these to specific events by referencing the
`.gtf` and `.bed` files, and XT tag from featureCountsReadAssignments. Next, the script uses the
XS tag from featureCountsReadAssignment to assign each read in the pair as the forward or reverse
read, relative to the direction of transcription. Finally, it looks for T>C in FW read, A>G in RV
read, checks these are not present in the SNP VCF file, and outputs metadata on each read-pair
about it's event assignment, number of conversions, coverage etc.
"""

# mypy: ignore-errors
# pylint: disable=fixme

import sys
from argparse import ArgumentParser
from pathlib import Path

import pandas as pd
import polars as pl

from isoslam import io, isoslam


def main(argv=None):
    """Script main.
    parses command line options in sys.argv, unless *argv* is given.
    """
    if argv is None:
        argv = sys.argv

    parser = ArgumentParser(description=__doc__)
    parser.add_argument("-c", "--config", dest="config_file", type=Path, help="Path to configuration file to load.")
    parser.add_argument(
        "-b",
        "--bam",
        dest="infile_bam",
        type=str,
        help="Supply a path to the bam file that has undergone read assignment with featureCounts",
    )
    parser.add_argument(
        "-g", "--gtf", dest="gtf_path", type=str, help="Supply a path to the transcript assembly gtf file"
    )
    parser.add_argument(
        "-bed", dest="utron_bed", type=str, help="Supply a path to the utron bed file. Must be bed6 format"
    )
    # TODO: Add option for locating vcf file
    parser.add_argument(
        "-o",
        "--out",
        dest="outfile_tsv",
        type=str,
        help="""Supply a path to the output file. This file will contain
                        conversions per pair, accounting for stranding""",
    )
    parser.add_argument("-vcf", "--vcf", dest="vcf_path", type=str, help="Supply a path to the VCF.gz file")
    parser.add_argument("--delim", dest="delim", type=str, help="Delimiter to use in output.")

    argv_as_dictionary = vars(argv)
    print(f"{argv_as_dictionary=}")

    # Load files...
    # bamfile = io.load_file(argv_as_dictionary["infile_bam"])
    vcffile = io.load_file(argv_as_dictionary["vcf_path"])
    # .bed file
    utron_coords = isoslam.extract_transcripts(argv_as_dictionary["utron_bed"])
    # .gtf file
    strand_dict, tx2gene = isoslam.extract_strand_transcript(argv_as_dictionary["gtf_path"])

    # Load configuration
    config = io.load_and_update_config(argv)

    def fragment_iterator(read_iterator):
        read_list = []
        last_read = None

        for read in read_iterator:
            if last_read is not None and last_read != read.query_name:
                yield read_list
                read_list = []
                last_read = read.query_name
            last_read = read.query_name
            read_list.append(read)

        yield read_list

    i_progress = 0
    i_total_progress = 0
    first_matched = 0
    i_output = 0
    delim = argv_as_dictionary["delim"]
    # Create Polars dataframe to hold results
    schema = {
        "read_uid": int,
        "transcript_id": str,
        "start": int,
        "end": int,
        "chr": str,
        "strand": str,
        "assignment": str,
        "conversions": int,
        "convertible": int,
        "coverage": int,
    }
    results = pl.DataFrame(schema=schema)
    with open(argv_as_dictionary["outfile_tsv"], "w", encoding="utf-8") as outfile:
        # Add column headers
        outfile.write(
            # TODO: Lowercase and 'chromosome' over 'Chr'
            f"Read_UID{delim}Transcript_id{delim}Start{delim}End{delim}Chr{delim}Strand{delim}"
            f"Assignment{delim}Conversions{delim}Convertible{delim}Coverage\n"
        )
        pd_results = pd.DataFrame()

        for pair in isoslam.extract_segment_pairs(argv_as_dictionary["infile_bam"]):
            if i_total_progress >= 2000000:
                break
            if first_matched > 1000:
                break

            if len(pair) != 2:
                continue

            read1, read2 = pair

            if read1.is_unmapped or read2.is_unmapped:
                continue

            i_total_progress += 1
            i_progress += 1

            if i_progress == 10000:
                i_progress = 0

            # Extract features -
            # TODO - It should be possible to modify extract_features_from_pair() to determine which is forward and
            #        which reverse in this initial call so that subsequently we have pair_features["forward"] and
            #        pair_features["reverse"] that we can work with.
            pair_features = isoslam.extract_features_from_pair(pair)

            # Use the intersection of sets, this skips if either read1 or read2 aren't assigned, - or +
            if not {"Assigned", "+", "-"} & {pair_features["read1"]["status"], pair_features["read2"]["status"]}:
                continue

            # pass if either is assigned
            status_list = [pair_features["read1"]["status"], pair_features["read2"]["status"]]
            # If both reads are assigned/+/-...
            if all(status in ["Assigned", "+", "-"] for status in status_list):
                # ...check if strands are equal if so they are assign one to strand
                if (
                    strand_dict[pair_features["read1"]["transcript"]]
                    == strand_dict[pair_features["read2"]["transcript"]]
                ):
                    strand = strand_dict[pair_features["read1"]["transcript"]]
                # ...if not we pass this pair
                else:
                    continue
            else:
                # If read1 is Assigned we get the ??? for this transcript, otherwise it comes from read2
                strand = (
                    strand_dict[pair_features["read1"]["transcript"]]
                    if pair_features["read1"]["status"] == "Assigned"
                    else strand_dict[pair_features["read2"]["transcript"]]
                )
            # Set forward and reverse reads
            # TODO: Extract is_reverse to the pair_features["read1"]["reverse"] in refactored code
            if read1.is_reverse and not read2.is_reverse:
                reverse_read = read1
                forward_read = read2
            elif read2.is_reverse and not read1.is_reverse:
                reverse_read = read2
                forward_read = read1
            else:
                # Not proper pair
                continue

            # Lists (utrons1 and utrons2) are iterated over further down
            pair_features["read1"]["utron"] = isoslam.extract_utron(
                features=pair_features["read1"], gene_transcript=tx2gene, coordinates=utron_coords
            )
            pair_features["read2"]["utron"] = isoslam.extract_utron(
                features=pair_features["read2"], gene_transcript=tx2gene, coordinates=utron_coords
            )

            ## @ns-rse 2023-12-20
            ## Get blocks to see how the reads are spliced
            ## Mainly use this to assign spliced reads
            ## But also use to check retained reads aren't spliced
            ## ALSO: There are some examples where a read would be called
            ## as retained because it overlaps an intron by 1 or 2 bases
            ## but these bases within the intron are the same as those after
            ## i.e. its ambiguous, but we can use evidence from the blocks of
            ## the partner read if it overlaps is spliced there.
            ## If not, we should cull assignment to those events
            ## We are comparing ret vs spliced, so instead of adding 1 to both
            ## we will just ignore those few
            block_starts1, block_ends1 = isoslam.zip_blocks(read1)
            block_starts2, block_ends2 = isoslam.zip_blocks(read2)

            # Build a dictionary as its cleaner to work with
            blocks = {
                "read1": {"starts": block_starts1, "ends": block_ends1},
                "read2": {"starts": block_starts2, "ends": block_ends2},
            }

            def print_info(chromosome, start, end, transcript_id, strand, blocks, text, count=i_total_progress) -> None:
                """Print information on progress to aid debugging."""
                print(f"\n{count} =========================== {text=}\n")
                print(f"{transcript_id=}")
                print(f"{chromosome=}")
                print(f"{start=}")
                print(f"{end=}")
                print(f"{strand=}")
                print(f"{blocks=}")

            # Retain within introns
            read1_within_intron = isoslam.filter_within_introns(pair_features, blocks, read="read1")
            read2_within_intron = isoslam.filter_within_introns(pair_features, blocks, read="read2")
            # Retain spliced
            read1_spliced_3UI = isoslam.filter_spliced_utrons(pair_features, blocks, read="read1")
            read2_spliced_3UI = isoslam.filter_spliced_utrons(pair_features, blocks, read="read2")

            all_dicts = [read1_within_intron, read2_within_intron, read1_spliced_3UI, read2_spliced_3UI]

            # Check that there are some retained regions (all dictionaries would be empty if there are none retained and
            # not {} evaluates to True, hence wrapping in all())
            if all(not contents for contents in all_dicts):
                continue
            first_matched += 1

            assign_conversions_to_retained = isoslam.unique_conversions(read1_within_intron, read2_within_intron)
            assign_conversions_to_spliced = isoslam.unique_conversions(read1_spliced_3UI, read2_spliced_3UI)

            ## If there are any events in both we want to remove them - this should be rare
            assign_conversions_to_retained, assign_conversions_to_spliced = isoslam.remove_common_reads(
                assign_conversions_to_retained, assign_conversions_to_spliced
            )

            # if we are mapped to a +ve stranded transcript, then count T>C in
            # the forward read and A>G in the reverse read.
            # TODO - Once we have assigned `read1/read2` to `forward/reverse` much earlier we can update this
            if strand == "+":
                coverage_counts = isoslam.count_conversions_across_pairs(
                    forward_read=forward_read,
                    reverse_read=reverse_read,
                    vcf_file=vcffile,
                    forward_conversion=config["forward_reads"],
                    reverse_conversion=config["reverse_reads"],
                )
            elif strand == "-":
                # If we are mapped to a -ve stranded transcript, count T>C in the reverse read and A>G
                # in the forward read.
                # can either flip the reads or the conversions
                coverage_counts = isoslam.count_conversions_across_pairs(
                    forward_read=reverse_read,
                    reverse_read=forward_read,
                    vcf_file=vcffile,
                    forward_conversion=config["forward_reads"],
                    reverse_conversion=config["reverse_reads"],
                )
            else:
                # should not be possible - but just in case
                pass

            i_output += 1

            # Stream output as a tsv
            # Format: read_uid, transcript_id, start, end, ret/spl, conversions, convertible, coverage
            io.write_assigned_conversions(
                assigned_conversions=assign_conversions_to_retained,
                coverage_counts=coverage_counts,
                read_uid=i_output,
                assignment="Ret",
                outfile=outfile,
                delim=argv_as_dictionary["delim"],
            )
            io.write_assigned_conversions(
                assigned_conversions=assign_conversions_to_spliced,
                coverage_counts=coverage_counts,
                read_uid=i_output,
                assignment="Spl",
                outfile=outfile,
                delim=argv_as_dictionary["delim"],
            )
            # A read pair will cover multiple lines if it matches multiple events (but metadata will be same)
            # ns-rse : Add in building Pandas dataframe so the function can return something that is testable
            results = isoslam.append_data(
                assigned_conversions=assign_conversions_to_retained,
                coverage_counts=coverage_counts,
                read_uid=i_output,
                assignment="Ret",
                results=results,
                schema=schema,
            )
            results = isoslam.append_data(
                assigned_conversions=assign_conversions_to_spliced,
                coverage_counts=coverage_counts,
                read_uid=i_output,
                assignment="Spl",
                results=results,
                schema=schema,
            )

    return results.sort(by=["read_uid", "transcript_id", "chr", "start", "end"])


if __name__ == "__main__":
    sys.exit(main(sys.argv))
