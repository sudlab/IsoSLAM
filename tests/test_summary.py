"""Tests of the summary module."""

from pathlib import Path
from re import Pattern
from typing import Any

import polars as pl
import pytest

from isoslam import summary

BASE_DIR = Path.cwd()
RESOURCES = BASE_DIR / "tests" / "resources"

# pylint: disable=too-many-arguments
# pylint: disable=too-many-positional-arguments
# pylint: disable=protected-access


@pytest.mark.parametrize(
    ("file_ext", "directory", "shape", "unique_filenames", "start_min", "start_max", "end_min", "end_max"),
    [
        pytest.param(
            ".tsv",
            RESOURCES / "results",
            (1696, 11),
            {"d0_0hr1_filtered_remapped_sorted", "d0_no4sU_filtered_remapped_sorted"},
            14940,
            25004,
            15080,
            29601,
            id="tsv",
        ),
        pytest.param(
            ".parquet",
            RESOURCES / "parquet",
            (7000, 11),
            {
                "d0_0hr1",
                "d0_0hr2",
                "d0_0hr3",
                "d0_0hr4",
                "d0_12hr1",
                "d0_12hr2",
                "d0_12hr3",
                "d0_12hr4",
                "d0_3hr1",
                "d0_3hr2",
                "d0_3hr3",
                "d0_3hr4",
                "d0_no4sU",
                "d16_0hr1",
                "d16_0hr2",
                "d16_0hr3",
                "d16_12hr1",
                "d16_12hr3",
                "d16_3hr1",
                "d16_3hr2",
                "d16_3hr3",
                "d16_no4sU",
                "d2_0hr1",
                "d2_0hr2",
                "d2_0hr3",
                "d2_0hr4",
                "d2_12hr1",
                "d2_12hr2",
                "d2_12hr3",
                "d2_12hr4",
                "d2_3hr1",
                "d2_3hr2",
                "d2_3hr3",
                "d2_3hr4",
                "d2_no4sU",
            },
            189339,
            244409112,
            189967,
            244409243,
            id="parquet",
        ),
    ],
)
def test_append_files(
    file_ext: str,
    directory: Path,
    shape: tuple[int, int],
    unique_filenames: list[str],
    start_min: int,
    start_max: int,
    end_min: int,
    end_max: int,
) -> None:
    """Test appending of files."""
    all_files = summary.append_files(file_ext, directory)
    assert isinstance(all_files, pl.DataFrame)
    assert all_files.shape == shape
    assert set(all_files["filename"].unique()) == unique_filenames
    assert all_files["Start"].min() == start_min
    assert all_files["Start"].max() == start_max
    assert all_files["End"].min() == end_min
    assert all_files["End"].max() == end_max


@pytest.mark.parametrize(
    ("file_ext", "directory", "regex", "expected_files", "expected_numbers"),
    [
        pytest.param(
            ".tsv",
            RESOURCES / "results",
            r"^d(\w+)_(\w+)hr(\w+)_",
            {"d0_0hr1_filtered_remapped_sorted"},
            {
                "count_max": 180,
                "count_min": 1,
                "total_max": 267,
                "total_min": 1,
                "percent_max": 1.0,
                "percent_min": 0.07142857142857142,
            },
            id="tsv",
        ),
        pytest.param(
            ".parquet",
            RESOURCES / "parquet",
            r"^d(\w+)_(\w+)hr(\w+)",
            {
                "d0_0hr1",
                "d0_0hr2",
                "d0_0hr3",
                "d0_0hr4",
                "d0_12hr1",
                "d0_12hr2",
                "d0_12hr3",
                "d0_12hr4",
                "d0_3hr1",
                "d0_3hr2",
                "d0_3hr3",
                "d0_3hr4",
                "d16_0hr1",
                "d16_0hr2",
                "d16_0hr3",
                "d16_12hr1",
                "d16_12hr3",
                "d16_3hr1",
                "d16_3hr2",
                "d16_3hr3",
                "d2_0hr1",
                "d2_0hr2",
                "d2_0hr3",
                "d2_0hr4",
                "d2_12hr1",
                "d2_12hr2",
                "d2_12hr3",
                "d2_12hr4",
                "d2_3hr1",
                "d2_3hr2",
                "d2_3hr3",
                "d2_3hr4",
            },
            {
                "count_max": 5,
                "count_min": 1,
                "total_max": 6,
                "total_min": 1,
                "percent_max": 1.0,
                "percent_min": 0.16666666666666666,
            },
            id="parquet",
        ),
    ],
)
def test_summary_counts(
    file_ext: str, directory: Path, regex: Pattern, expected_files: set, expected_numbers, regtest
) -> None:
    """Test summary counts are correctly calculated."""
    summary_counts = summary.summary_counts(file_ext, directory, regex=regex)
    assert isinstance(summary_counts, pl.DataFrame)
    assert set(summary_counts["filename"].unique()) == expected_files
    assert summary_counts["conversion_count"].max() == expected_numbers["count_max"]
    assert summary_counts["conversion_count"].min() == expected_numbers["count_min"]
    assert summary_counts["conversion_total"].max() == expected_numbers["total_max"]
    assert summary_counts["conversion_total"].min() == expected_numbers["total_min"]
    assert summary_counts["conversion_percent"].max() == expected_numbers["percent_max"]
    assert summary_counts["conversion_percent"].min() == expected_numbers["percent_min"]
    print(summary_counts.write_csv(), file=regtest)


@pytest.mark.parametrize(
    ("df", "column", "regex"),
    [
        pytest.param(
            pl.DataFrame(
                {
                    "filename": [
                        "d0_0hr1",
                        "d0_0hr2",
                        "d0_0hr3",
                        "d0_0hr4",
                        "d0_12hr1",
                        "d0_12hr2",
                        "d0_12hr3",
                        "d0_12hr4",
                        "d0_3hr1",
                        "d0_3hr2",
                        "d0_3hr3",
                        "d0_3hr4",
                        "d0_no4sU",
                        "d16_0hr1",
                        "d16_0hr2",
                        "d16_0hr3",
                        "d16_12hr1",
                        "d16_12hr3",
                    ]
                }
            ),
            "filename",
            r"^d(\w+)_(\w+)hr(\w+)",
            id="simple",
        ),
        pytest.param(
            pl.DataFrame(
                {
                    "filename": [
                        "d0_0hr1_EKRN230046546-1A_HFWGNDSX7_L2",
                        "d0_0hr2_EKRN230046547-1A_HFWGNDSX7_L2",
                        "d0_0hr3_EKRN230046548-1A_HFWGNDSX7_L2",
                        "d0_0hr4_EKRN230046549-1A_HFWGNDSX7_L2",
                        "d0_12hr1_EKRN230046554-1A_HFWGNDSX7_L2",
                        "d0_12hr2_EKRN230046555-1A_HFWGNDSX7_L2",
                        "d0_12hr3_EKRN230046556-1A_HFWGNDSX7_L2",
                        "d0_12hr4_EKRN230046557-1A_HFWGNDSX7_L2",
                        "d0_no4sU_EKRN230046554-1A_HFWGNDSX7_L2",
                    ]
                }
            ),
            "filename",
            r"^d(\w+)_(\w+)hr(\w+)_",
            id="realistic",
        ),
    ],
)
def test_extract_hour_and_replicate(df: pl.DataFrame, column: str, regex: Pattern, regtest) -> None:
    """Test extraction of hour and replicate from filename."""
    summary_counts = summary.extract_day_hour_and_replicate(df, column, regex)
    print(summary_counts.write_csv(), file=regtest)


@pytest.mark.parametrize(
    ("df", "groupby", "converted"),
    [
        pytest.param(
            pl.DataFrame(
                {
                    "transcript_id": ["a", "a", "b", "b", "c", "d", "d", "e"],
                    "chr": ["chr1", "chr1", "chr2", "chr2", "chr3", "chr4", "chr4", "chr5"],
                    "conversion": [True, False, True, False, False, False, True, True],
                    "n": [1, 2, 3, 4, 5, 6, 7, 8],
                }
            ),
            ["transcript_id", "chr"],
            "conversion",
            id="missing conversion True for transcript_id on chr 3",
        ),
        pytest.param(
            "sample_data_summary_counts",
            [
                "Transcript_id",
                "Strand",
                "Start",
                "End",
                "Assignment",
                "day",
                "hour",
                "replicate",
            ],
            "one_or_more_conversion",
            id="real data",
        ),
    ],
)
def test_aggregate_conversions(
    df: pl.DataFrame | str,
    groupby: list[str],
    converted: str,
    request: pytest.FixtureRequest,
    regtest,
) -> None:
    """Test derivation on non-captured dataset."""
    df = request.getfixturevalue(df) if isinstance(df, str) else df
    aggregated_conversions = summary._aggregate_conversions(df, groupby, converted)
    print(aggregated_conversions.write_csv(), file=regtest)


@pytest.mark.parametrize(
    ("df", "groupby", "converted"),
    [
        pytest.param(
            pl.DataFrame(
                {
                    "transcript_id": ["a", "b", "c", "d", "e"],
                    "chr": ["chr1", "chr2", "chr3", "chr4", "chr5"],
                    "len": [2, 2, 1, 2, 1],
                    "conversion": [True, True, False, True, True],
                },
                schema={
                    "transcript_id": pl.datatypes.String,
                    "chr": pl.datatypes.String,
                    "len": pl.datatypes.UInt32,
                    "conversion": pl.datatypes.Boolean,
                },
            ),
            ["transcript_id", "chr"],
            "conversion",
            id="missing conversion True for transcript_id on chr 3",
        ),
        pytest.param(
            "sample_data_summary_counts",
            None,
            "one_or_more_conversion",
            id="real data",
        ),
    ],
)
def test_filter_no_conversions(
    df: pl.DataFrame | str,
    groupby: list[str],
    converted: str,
    request: pytest.FixtureRequest,
    regtest,
) -> None:
    """Test filtering of non conversions."""
    df = request.getfixturevalue(df) if isinstance(df, str) else df
    filtered_no_conversions = summary._filter_no_conversions(df, groupby, converted)
    print(filtered_no_conversions.write_csv(), file=regtest)


@pytest.mark.parametrize(
    ("df", "groupby", "converted"),
    [
        pytest.param(
            "sample_data_summary_counts",
            None,
            "one_or_more_conversion",
            id="real data groupby none",
        ),
    ],
)
def test_get_one_or_more_conversion(
    df: pl.DataFrame, groupby: list[str], converted: str, request: pytest.FixtureRequest, regtest
) -> None:
    """Test that inner_join_no_conversions returns the correct subset."""
    df = request.getfixturevalue(df) if isinstance(df, str) else df
    one_or_more_conversions = summary._get_one_or_more_conversion(df, groupby, converted)
    print(one_or_more_conversions.write_csv(), file=regtest)


@pytest.mark.parametrize(
    ("df", "groupby", "count", "total"),
    [
        pytest.param(
            pl.DataFrame(
                {
                    "day": [0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1],
                    "hour": [0, 0, 0, 0, 8, 8, 8, 8, 0, 0, 0, 0, 16, 16, 16, 16],
                    "replicate": [1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4],
                    "count": [2, 2, 2, 2, 1, 1, 1, 1, 5, 10, 15, 20, 0, 0, 0, 0],
                    "total": [8, 8, 8, 8, 1, 2, 3, 4, 10, 20, 30, 40, 1, 1, 1, 1],
                    "percent": [0.25, 0.25, 0.25, 0.24, 1.0, 0.5, 0.333333333, 0.25, 0.5, 0.5, 0.5, 0.5, 0, 0, 0, 0],
                }
            ),
            ["day", "hour"],
            "count",
            "total",
            id="simple mean",
        ),
        pytest.param(
            "one_or_more_conversions",
            None,
            "conversion_count",
            "conversion_total",
            id="real data mean",
        ),
    ],
)
def test_percent_conversions_across_replicates(
    df: pl.DataFrame | str, groupby: list[str], count: str, total: str, request: pytest.FixtureRequest, regtest
) -> None:
    """Test the summary._average_replicate() function."""
    df = request.getfixturevalue(df) if isinstance(df, str) else df
    df_average = summary._percent_conversions_across_replicates(df, groupby, count, total)
    print(df_average.write_csv(), file=regtest)


@pytest.mark.parametrize(
    ("df", "base_day", "base_hour"),
    [
        pytest.param(
            pl.DataFrame(
                {
                    "transcript_id": "1A",
                    "day": [0, 0, 1, 1],
                    "hour": [0, 8, 0, 16],
                    "conversion_count": [8, 4, 50, 0],
                    "conversion_total": [32, 10, 100, 4],
                    "conversion_percent": [25.0, 40.0, 50.0, 0.0],
                }
            ),
            0,
            0,
            id="simple",
        ),
        pytest.param("averaged_data", 0, 0, id="real"),
    ],
)
def test_select_base_levels(
    df: pl.DataFrame | str,
    base_day: int,
    base_hour: int,
    request: pytest.FixtureRequest,
    regtest,
) -> None:
    """Test subsetting of data for baseline values."""
    df = request.getfixturevalue(df) if isinstance(df, str) else df
    baseline = summary._select_base_levels(df, base_day, base_hour)
    print(baseline.write_csv(), file=regtest)


@pytest.mark.parametrize(
    ("df_average", "df_baseline", "join_on"),
    [
        pytest.param(
            pl.DataFrame(
                {
                    "transcript_id": "1A",
                    "day": [0, 0, 1, 1],
                    "hour": [0, 8, 0, 16],
                    "conversion_count": [8, 4, 50, 0],
                    "conversion_total": [32, 10, 100, 4],
                    "conversion_percent": [25.0, 40.0, 50.0, 0.0],
                }
            ),
            pl.DataFrame(
                {
                    "transcript_id": "1A",
                    "baseline_count": [8.0],
                    "baseline_total": [32.0],
                    "baseline_percent": [25.0],
                }
            ),
            ["transcript_id"],
            id="simple",
        ),
        pytest.param(
            "averaged_data", "baseline_mean", ["Transcript_id", "Strand", "Start", "End", "Assignment"], id="real"
        ),
    ],
)
def test_merge_average_baseline(
    df_average: pl.DataFrame | str,
    df_baseline: pl.DataFrame | str,
    join_on: list[str],
    request: pytest.FixtureRequest,
    regtest,
) -> None:
    """Test merging of average and baseline data."""
    df_average = request.getfixturevalue(df_average) if isinstance(df_average, str) else df_average
    df_baseline = request.getfixturevalue(df_baseline) if isinstance(df_baseline, str) else df_baseline
    combined = summary._merge_average_with_baseline(df_average, df_baseline, join_on)
    print(combined.write_csv(), file=regtest)


@pytest.mark.parametrize(
    ("df", "to_normalise", "baseline"),
    [
        pytest.param(
            pl.DataFrame(
                {
                    "transcript_id": ["1A", "1A", "1A", "1A"],
                    "day": [0, 0, 1, 1],
                    "hour": [0, 8, 0, 16],
                    "convertion_count": [8, 4, 50, 0],
                    "conversion_total": [32, 10, 100, 4],
                    "conversion_percent": [25.0, 40.0, 50.0, 0.0],
                    "baselin_count": [8.0, 8.0, 8.0, 8.0],
                    "baseline_total": [32.0, 32.0, 32.0, 32.0],
                    "baseline_percent": [25.0, 25.0, 25.0, 25.0],
                }
            ),
            "conversion_percent",
            "baseline_percent",
            id="simple",
        ),
        pytest.param("merged_average_baseline", "conversion_percent", "baseline_percent", id="real"),
    ],
)
def test_normalise(
    df: pl.DataFrame | str, to_normalise: str, baseline: str, request: pytest.FixtureRequest, regtest
) -> None:
    """Test the summary._normalise() function divides the target (''to_normalise'') by ''baseline''."""
    df = request.getfixturevalue(df) if isinstance(df, str) else df
    normalised = summary._normalise(df, to_normalise, baseline)
    print(normalised.write_csv(), file=regtest)


@pytest.mark.parametrize(
    (
        "file_ext",
        "directory",
        "groupby",
        "conversions_var",
        "conversions_threshold",
        "test_file",
        "regex",
        "expected_numbers",
    ),
    [
        pytest.param(
            ".tsv",
            RESOURCES / "tsv",
            [
                "Transcript_id",
                "Start",
                "End",
                "Chr",
                "Strand",
                "Assignment",
                "filename",
            ],
            "Conversions",
            1,
            "no4sU",
            r"^d(\w+)_(\w+)hr(\w+)",
            {
                "shape": (6108, 14),
                "count_max": 5,
                "count_min": 1,
                "total_max": 6,
                "total_min": 1,
                "percent_max": 1.0,
                "percent_min": 0.16666666666666666,
                "unique_files": 32,
            },
            id="tsv",
        ),
        pytest.param(
            ".parquet",
            RESOURCES / "parquet",
            [
                "Transcript_id",
                "Start",
                "End",
                "Chr",
                "Strand",
                "Assignment",
                "filename",
            ],
            "Conversions",
            1,
            "no4sU",
            r"^d(\w+)_(\w+)hr(\w+)",
            {
                "shape": (6108, 14),
                "count_max": 5,
                "count_min": 1,
                "total_max": 6,
                "total_min": 1,
                "percent_max": 1.0,
                "percent_min": 0.16666666666666666,
                "unique_files": 32,
            },
            id="parquet",
        ),
    ],
)
def test_statistics_class(
    file_ext: str,
    directory: str,
    groupby: list[str],
    conversions_var: str,
    conversions_threshold: int,
    test_file: str,
    regex: Pattern,
    expected_numbers: dict[str, Any],
) -> None:
    """Test instantiation of statistics class."""
    statistics = summary.Statistics(
        file_ext=file_ext,
        directory=directory,
        groupby=groupby,
        conversions_var=conversions_var,
        conversions_threshold=conversions_threshold,
        test_file=test_file,
        regex=regex,
    )
    assert statistics.file_ext == file_ext
    assert statistics.directory == directory
    assert statistics.regex == regex
    assert statistics.groupby == groupby
    assert statistics.conversions_var == conversions_var
    assert statistics.conversions_threshold == conversions_threshold
    assert statistics.test_file == test_file
    # Check the data frame
    assert isinstance(statistics.data, pl.DataFrame)
    assert statistics.shape == expected_numbers["shape"]
    assert statistics.data["conversion_count"].max() == expected_numbers["count_max"]
    assert statistics.data["conversion_count"].min() == expected_numbers["count_min"]
    assert statistics.data["conversion_total"].max() == expected_numbers["total_max"]
    assert statistics.data["conversion_total"].min() == expected_numbers["total_min"]
    assert statistics.data["conversion_percent"].max() == expected_numbers["percent_max"]
    assert statistics.data["conversion_percent"].min() == expected_numbers["percent_min"]
    assert statistics.unique == expected_numbers["unique_files"]
