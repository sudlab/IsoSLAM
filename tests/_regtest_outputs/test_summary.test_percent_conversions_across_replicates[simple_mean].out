day,hour,count,total,conversion_percent
0,0,8,32,25.0
0,8,4,10,40.0
1,0,50,100,50.0
1,16,0,4,0.0

